File: frontend/src/api/students.ts
import { apiFetch } from "./client";

export interface StudentBasic {
  student_id: number;
  name: string;
  school: string | null;
  grade: number | null;
  class_name: string | null;
  note: string | null;
}

export interface PrismScore {
  eval_date: string;
  score_p: number;
  score_r: number;
  score_i: number;
  score_s: number;
  score_m: number;
  note: string | null;
}

export interface ClinicErrorCount {
  code: string;
  label: string;
  count: number;
  category: "이해" | "수행";
}

export interface SelfPredictEntry {
  log_date: string;
  title: string | null;
  predicted: number | null;
  actual: number | null;
  gap: number;
  note: string | null;
}

export interface SelfPredictChartPoint {
  date: string;
  predicted: number | null;
  actual: number | null;
}

export interface ProgressEntry {
  log_date: string;
  category: string;
  chapter: string | null;
  title: string | null;
  planned: string | null;
  actual: string | null;
  score_display: string;
  note: string | null;
}

export interface QuantityGradeCount {
  grade: string;
  count: number;
  pct: number | null;
}

export interface QualitativeEntry {
  eval_date: string;
  note_completion: number;
  written_completion: number;
  textbook_marking: number;
  second_solve_reason: number;
  note: string | null;
}

export interface LogEntry {
  log_date: string;
  log_type: string;
  summary: string | null;
  detail: string | null;
}

export interface StudentDashboard {
  student: StudentBasic;
  clinic_count: number;
  latest_prism: PrismScore | null;
  avg_ris: number | null;
  avg_pm: number | null;
  prism_history: PrismScore[];
  clinic_error_distribution: ClinicErrorCount[];
  self_predict_entries: SelfPredictEntry[];
  self_predict_chart: SelfPredictChartPoint[];
  self_predict_avg_gap: number | null;
  self_predict_over_count: number;
  self_predict_under_count: number;
  progress_entries: ProgressEntry[];
  quantity_grade_counts: QuantityGradeCount[];
  latest_qualitative: QualitativeEntry | null;
  logs: LogEntry[];
}

export interface StudentTemplateItem {
  filename: string;
  label: string;
  available: boolean;
  url: string | null;
}

export function fetchStudents(): Promise<StudentBasic[]> {
  return apiFetch<StudentBasic[]>("/api/students");
}

export function fetchStudentTemplates(): Promise<{ items: StudentTemplateItem[] }> {
  return apiFetch("/api/students/templates");
}

export function fetchStudentDashboard(sid: number): Promise<StudentDashboard> {
  return apiFetch<StudentDashboard>(`/api/students/${sid}/dashboard`);
}

export function updateStudentBasic(
  sid: number,
  body: { school: string | null; grade: number; class_name: string | null; note: string | null },
): Promise<StudentDashboard> {
  return apiFetch<StudentDashboard>(`/api/students/${sid}`, {
    method: "PUT",
    body: JSON.stringify(body),
  });
}

export function addPrismEval(
  sid: number,
  body: {
    eval_date: string; score_p: number; score_r: number; score_i: number;
    score_s: number; score_m: number; note: string | null;
  },
): Promise<StudentDashboard> {
  return apiFetch<StudentDashboard>(`/api/students/${sid}/prism`, {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export function addSelfPredict(
  sid: number,
  body: { log_date: string; title: string | null; predicted: number; actual: number; note: string | null },
): Promise<StudentDashboard> {
  return apiFetch<StudentDashboard>(`/api/students/${sid}/self-predict`, {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export function addProgress(
  sid: number,
  body: {
    log_date: string; category: "진도" | "숙제" | "시험"; chapter: string | null;
    title: string | null; planned: string | null; actual: string | null;
    score_raw: number | null; score_max: number | null; note: string | null;
  },
): Promise<StudentDashboard> {
  return apiFetch<StudentDashboard>(`/api/students/${sid}/progress`, {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export function addQuantityAssessment(
  sid: number,
  body: { eval_date: string; grade: "A" | "B" | "C" | "D"; note: string | null },
): Promise<StudentDashboard> {
  return apiFetch<StudentDashboard>(`/api/students/${sid}/assessment/quantity`, {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export function addQualitativeAssessment(
  sid: number,
  body: {
    eval_date: string; note_completion: number; written_completion: number;
    textbook_marking: number; second_solve_reason: number; note: string | null;
  },
): Promise<StudentDashboard> {
  return apiFetch<StudentDashboard>(`/api/students/${sid}/assessment/qualitative`, {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export function addManagementLog(
  sid: number,
  body: { log_date: string; log_type: "보호자" | "출결" | "메모"; summary: string | null; detail: string | null },
): Promise<StudentDashboard> {
  return apiFetch<StudentDashboard>(`/api/students/${sid}/logs`, {
    method: "POST",
    body: JSON.stringify(body),
  });
}

File: backend/app/schemas/students.py
from __future__ import annotations

from typing import Literal, Optional

from pydantic import BaseModel, Field


class StudentBasic(BaseModel):
    student_id: int
    name: str
    school: Optional[str] = None
    grade: Optional[int] = None
    class_name: Optional[str] = None
    note: Optional[str] = None


class CreateStudentRequest(BaseModel):
    name: str
    school: Optional[str] = None
    grade: int = Field(default=1, ge=1, le=6)
    class_name: Optional[str] = None


class PrismScore(BaseModel):
    eval_date: str
    score_p: int
    score_r: int
    score_i: int
    score_s: int
    score_m: int
    note: Optional[str] = None


class ClinicErrorCount(BaseModel):
    code: str
    label: str
    count: int
    category: Literal["이해", "수행"]


class SelfPredictEntry(BaseModel):
    log_date: str
    title: Optional[str] = None
    predicted: Optional[int] = None
    actual: Optional[int] = None
    gap: int
    note: Optional[str] = None


class ProgressEntry(BaseModel):
    log_date: str
    category: str
    chapter: Optional[str] = None
    title: Optional[str] = None
    planned: Optional[str] = None
    actual: Optional[str] = None
    score_display: str
    note: Optional[str] = None


class QuantityGradeCount(BaseModel):
    grade: str
    count: int
    pct: Optional[float] = None


class QualitativeEntry(BaseModel):
    eval_date: str
    note_completion: int
    written_completion: int
    textbook_marking: int
    second_solve_reason: int
    note: Optional[str] = None


class LogEntry(BaseModel):
    log_date: str
    log_type: str
    summary: Optional[str] = None
    detail: Optional[str] = None


class StudentDashboard(BaseModel):
    student: StudentBasic
    clinic_count: int
    latest_prism: Optional[PrismScore] = None
    avg_ris: Optional[float] = None
    avg_pm: Optional[float] = None
    prism_history: list[PrismScore] = []
    clinic_error_distribution: list[ClinicErrorCount] = []
    self_predict_entries: list[SelfPredictEntry] = []
    self_predict_chart: list[dict] = []  # [{date, predicted, actual}] 일자별 평균 집약
    self_predict_avg_gap: Optional[float] = None
    self_predict_over_count: int = 0
    self_predict_under_count: int = 0
    progress_entries: list[ProgressEntry] = []
    quantity_grade_counts: list[QuantityGradeCount] = []
    latest_qualitative: Optional[QualitativeEntry] = None
    logs: list[LogEntry] = []


class UpdateStudentBasicRequest(BaseModel):
    school: Optional[str] = None
    grade: int = Field(ge=1, le=6)
    class_name: Optional[str] = None
    note: Optional[str] = None


class PrismEvalRequest(BaseModel):
    eval_date: str
    score_p: int = Field(ge=1, le=5)
    score_r: int = Field(ge=1, le=5)
    score_i: int = Field(ge=1, le=5)
    score_s: int = Field(ge=1, le=5)
    score_m: int = Field(ge=1, le=5)
    note: Optional[str] = None


class SelfPredictRequest(BaseModel):
    log_date: str
    title: Optional[str] = None
    predicted: int = Field(ge=0, le=100)
    actual: int = Field(ge=0, le=100)
    note: Optional[str] = None


class ProgressLogRequest(BaseModel):
    log_date: str
    category: Literal["진도", "숙제", "시험"]
    chapter: Optional[str] = None
    title: Optional[str] = None
    planned: Optional[str] = None
    actual: Optional[str] = None
    score_raw: Optional[int] = None
    score_max: Optional[int] = None
    note: Optional[str] = None


class QuantityAssessmentRequest(BaseModel):
    eval_date: str
    grade: Literal["A", "B", "C", "D"]
    note: Optional[str] = None


class QualitativeAssessmentRequest(BaseModel):
    eval_date: str
    note_completion: int = Field(ge=1, le=5)
    written_completion: int = Field(ge=1, le=5)
    textbook_marking: int = Field(ge=1, le=5)
    second_solve_reason: int = Field(ge=1, le=5)
    note: Optional[str] = None


class ManagementLogRequest(BaseModel):
    log_date: str
    log_type: Literal["보호자", "출결", "메모"]
    summary: Optional[str] = None
    detail: Optional[str] = None


class StudentTemplateItem(BaseModel):
    filename: str
    label: str
    available: bool
    url: Optional[str] = None


class StudentTemplatesResponse(BaseModel):
    items: list[StudentTemplateItem]
